### ISSUE TITLE
1.0.5 breaks data merge

### ISSUE THREAD

**Issue by nicooprat** [vuejs/vue-test-utils#1678]
### Subject of the issue

Upgrading from 1.0.4 to 1.0.5 makes some of our tests to fail. Looks like the data merge is broken, and adding a `console.log()` in the `data()` function at wrapper creation is now displayed twice:

With 1.0.4:

![image](https://user-images.githubusercontent.com/645641/91986579-26e3dc80-ed2d-11ea-92cf-3648b36d7abe.png)

With 1.0.5:

![image](https://user-images.githubusercontent.com/645641/91982484-973d2e80-ed2a-11ea-8eff-e0c70f4f7923.png)

The first one pass through here: https://github.com/vuejs/vue/blob/7912f75c5eb09e0aef3e4bfd8a3bb78cad7540d7/src/core/util/options.js#L104

The second time through there: https://github.com/vuejs/vue/blob/7912f75c5eb09e0aef3e4bfd8a3bb78cad7540d7/src/core/util/options.js#L98

Our code creates a wrapper by merging instance options like this:

`const wrapper = shallowMount(OpportunityFeedback, { data() { ... });`

### Steps to reproduce

Tell us how to reproduce this issue. Please provide a working and simplified example.

### Expected behaviour

The data gets merged correctly as before.

### Actual behaviour

Data become undefined and this errocr is printed:

`[Vue warn]: Avoid adding reactive properties to a Vue instance or its root $data at runtime - declare it upfront in the data option.`

### Possible Solution

Sticking to 1.0.4 for now. Maybe related? https://github.com/vuejs/vue-test-utils/issues/1677

**Comment by afontcu** [comment]
Here's the diff between 1.0.4 and 1.0.5: https://github.com/vuejs/vue-test-utils/compare/v1.0.4...v1.0.5

Not sure how, but these are the only lines I could see as potentially related? https://github.com/vuejs/vue-test-utils/compare/v1.0.4...v1.0.5#diff-68ebb4dd9f4503e2b26a1b1ac98d4fa2R127-R132 (https://github.com/vuejs/vue-test-utils/pull/1661)

**Comment by nicooprat** [comment]
You're right, I just tried to edit this file `node_modules/@vue/test-utils/dist/vue-test-utils.js` and rollbacked the changes you mentioned, and it works again...

**Comment by nicooprat** [comment]
So it looks like this feature is broken: https://vue-test-utils.vuejs.org/api/options.html#data

For now I removed the errors by using [`await setData({ ... })`](https://vue-test-utils.vuejs.org/api/wrapper/#setdata) instead of merging.

**Comment by lmiller1990** [comment]
Uh oh, this is a pretty major bug. I will look into this. Sorry I missed this one.

Can you post a reproduction? I'll try to reproduce in the meantime.

**Comment by lmiller1990** [comment]
I cannot reproduce (or, I don't understand how?). Can you provide a reproduction I can run locally?

I made this component:

```html
<template>
  <div class="hello">
    <h1>{{ msg }}</h1>
  </div>
</template>

<script lang="ts">
import Vue from 'vue';

export default Vue.extend({
  name: 'HelloWorld',
  data() {
    return {
      foo: 'bar',
      msg: 'nooo'
    }
  }
});
</script>
```

Test: 

```js
import { mount } from '@vue/test-utils'
import HelloWorld from '@/components/HelloWorld.vue'

describe('HelloWorld.vue', () => {
  it('test one', () => {
    const msg = 'ok'

    const wrapper = mount(HelloWorld, {
      data() {
        return {
          msg: 'ok'
        }
      }
    })

    expect(wrapper.text()).toMatch(msg)
  })
})
```

And it's all good.

Most likely this was introduced by the feature added to make `ctx.$store` etc available with the composition AP plugin. But I will need to reproduce before I can fix it. Thanks!

**Comment by AleksandrSl** [comment]
Hi, seems that I encountered similar bug.
On 1.0.4 everything works fine, but fails on  > 1.0.5

The most simple reproduction I come up wtih:
```js
import { createLocalVue, shallowMount } from "@vue/test-utils"
import GenericAnalysesReview from "@/components/Bug.vue"


const localVue = createLocalVue()

async function stubComponent(
    data
) {
    return shallowMount(GenericAnalysesReview, {
        localVue,
        data() {
            return data
        },
    })
}

describe("Some test", () => {
    it("will fail", async () => {
        await stubComponent({})
    })
})
```

"@/components/Bug.vue"
```vue
<template>
    <div v-show="runs.length === 1"></div>
</template>

<script>
    export default {
        name: "GenericAnalysesReview",
        data() {
            return {
                runs: [],
            }
        }
    }
</script>
```


error
```
[Vue warn]: Avoid adding reactive properties to a Vue instance or its root $data at runtime - declare it upfront in the data option.
[Vue warn]: Property or method "runs" is not defined on the instance but referenced during render. Make sure that this property is reactive, either in the data option, or for class-based components, by initializing the property. See: https://vuejs.org/v2/guide/reactivity.html#Declaring-Reactive-Properties.

found in

---> <GenericAnalysesReview> at src/components/Bug.vue
       <Root>
[Vue warn]: Error in render: "TypeError: Cannot read property 'length' of undefined"

found in

---> <GenericAnalysesReview> at src/components/Bug.vue
       <Root>
TypeError: Cannot read property 'length' of undefined ...
```


One weird notion, when `{}` is used directly, everything works fine (maybe some transpilation/optimization magic in this case) e.g. 
```js
import { createLocalVue, shallowMount } from "@vue/test-utils"
import GenericAnalysesReview from "@/components/Bug.vue"


const localVue = createLocalVue()

async function stubComponent(
    data
) {
    return shallowMount(GenericAnalysesReview, {
        localVue,
        data() {
            return data
        },
    })
}

describe("Some test", () => {

    it("will fail", async () => {
        await stubComponent({})
    })

    it("will not fail", async () => {
        shallowMount(GenericAnalysesReview, {
            localVue,
            data() {
                return {}
            },
        })
    })
})

```

**Comment by lmiller1990** [comment]
Thanks for the repro. I will try looking at this soon.

95% chance this commit is causing the problem https://github.com/vuejs/vue-test-utils/commit/f78f817306e2e228319092e3d3de55fc4356aa93

So if anyone else is motivated or has time you can probably try reverting this and see if that fixes the problem, should be good first step to solving this.

**Comment by nicooprat** [comment]
Thanks for the repro @AleksandrSl . Indeed I encounter this issue only when using a function to generate generic component overrides. 

> So if anyone else is motivated or has time you can probably try reverting this and see if that fixes the problem, 

I can confirm that reverting it solves the issue. Couldn't find a working alternative for now though...
